# Overview of Integers and Numbers
We already know the following operators which may be applied to numbers: +, -, * and \*\*. The division operator / for integers gives a floating-point real number (an object of type float). The exponentiation ** also returns a float when the power is negative:

In [1]:
print(17 / 3) # gives 5.66666666667
print(2 ** 4)  # gives 16
print(2 ** -2) # gives 0.25

5.666666666666667
16
0.25


* There's a special operation for integer division where the remainder is discarded: //.
* The operation that yields a remainder of such a division looks like %. Both operation always yield an object of type int.

In [2]:
print(17 / 3)  # gives 5.66666666667
print(17 // 3) # gives 5
print(17 % 3) # gives 2

5.666666666666667
5
2


* The **int()** function drops the decimal portion of a floating point number without rounding:

In [ ]:
print(int(1.3))   # gives 1
print(int(1.7))   # gives 1
print(int(-1.3))  # gives -1
print(int(-1.7))  # gives -1

* There's also a function **round()** that performs the usual rounding:

In [ ]:
print(round(1.3)) # gives 1
print(round(1.7)) # gives 2
print(round(-1.3)) # gives -1
print(round(-1.7)) # gives -2

* You can also specify the number of decimal digits to round to in the **round()** function:

In [ ]:
print(round(3.14159,2)) # gives 3.14
print(round(3.14159,3))  # gives 3.142
print(round(3.14159,4)) # gives 3.1416

# Integer Arithmetic
We already know the following operations which may be applied to numbers: +, -, * and **.
* Division / always gives you a float.
* Exponentiation ** also returns a float when the power is negative.

In [6]:
print(17 / 3)
print(20 / 4)
print(2 ** 4)
print(2 ** -3)

5.666666666666667
5.0
16
0.125


# Integer Division
* Division of an integer by another integer will always yield a floating, even if it is evenly divisible.
* The remainder operator, %, yields an integer that is the remainder after integer division.
The floor division operator, //, yield an integer that is portion of the division where the remainder is discarded.
* Remember that division by zero is undefined. Python will give you an error and your program will stop.

In [7]:
print(18 / 3) # Results in a float
print(17 / 3) # Results in a float
print(17 // 3) # Results in an int
print(17 % 3) # Results in an int

6.0
5.666666666666667
5
2


# Floating Point Numbers
* When you read an integer value, use input() and then cast a string to integer using int().
* When you read a floating-point number, you will cast a string to float using float().
    * Note that if you enter an integer value and then float it, the value becomes a floating point number. You know this because the value is display with a ".0" a the end.


# Very Big and Very Small Numbers
* You can write floats with very big or very small absolute values using the scientific notation.
*  The distance from the Earth to the Sun is $1.49 \times 10^{11}$  meters, or 1.496e11 in Python.
*  The mass of one molecule of water is $2.99 \times 10^{-23}$ grams, or 2.99e-23 in Python.
*  Note that Python will display only about 16 digits


In [8]:
from math import pi

EarthtoSun = 1.496e11 # Distance Earth to Sun in meters
print(EarthtoSun)

MassofWater = 2.99e-23 # Mass of a Molecule of MassofWater in grams
print (MassofWater)

LifeOnEarth = 3.9e9 # Life on Earth Began in Years
print(LifeOnEarth)

print(pi)

149600000000.0
2.99e-23
3900000000.0
3.141592653589793


# Difference between Round and Int
* You can also cast a float value to an int. Of course, it can be achieved using the int() function, which simply discards the fractional part (rounding, thus, every number towards 0).
* But, generally, it is much more convenient to use round(), which performs that rounding we are used to.
* You can also specify the number of digits past the decimal point when to round a number, as in:

In [ ]:
x = 12.455
print(round(x,2))

print(int(1.3))  # gives 1
print(int(1.7)) # gives 1
print(int(-1.3)) # gives -1
print(int(-1.7)). # gives -1

print(round(1.3)) # gives 1
print(round(1.7))  # gives 2
print(round(-1.3))  # gives -1
print(round(-1.7)) # gives -2

print(round(12.455,2))

# Floating Points are not Exact
* Floating-point real numbers can't be represented with exact precision due to hardware limitations. This can lead to some interesting effects.

In [ ]:
print(0.1 + 0.2) # gives 0.30000000000000004

# Math Module

* Python has many functions for calculations with floats. Most of them can be found in the math module. To use this module, you need to import it first by writing the following instruction at the beginning of your program:




In [ ]:
import math

* For example, if we want to find the ceiling value for x:  the smallest integer not less than x
    * we call the corresponding function from the math module: **math.ceil(x)**.
* You may also import only one method from the whole module and use it under its short name.


In [ ]:
from math import floor

* Now you can invoke the floor function by it's name: **floor(x)** <br>



In [ ]:
# We have to use math.ceil because we imported the entire math module
import math
x = 7/2
print(x)
y = math.ceil(x)
print(y)


# Here,  we can use floor because we explicitly imported it from the math module
from math import floor
x = 7/2
print(x)
y = floor(x)
print(y)

## Implicit Math Functions

Operator  | Name                        
-------------------|------------------  
int(x)      | returns the integer equivalent of x   
float(x) |  returns the floating equivalent of x  
round(x) | returns an integer rounded to the nearest whole number of x
abs(x)  | returns the absolute value of x



## Math Module Functions

Syntax | Function
-------------- |-------------------
floor(x)  | returns the largest integer less than or equal to x
ceil(x) | returns the smallest integer greater than or equal to x
sqrt(x)  | returns the square root of x
log(x) |  returns the natural logarithm of x
log(x,a) | returns the logarithm of x to base a
pi | mathematical constant pi = 3.1415926...
e | mathematical constant e = 2.71828...
sin(x) | returns the sine of x in radians
cos(x) | returns the cosine of x in radians
tan(x) | returns the tangent function of x
asin(x) | returns the arcsine of x in radians


# Formatting Values
* So far, we have printed the default format for values.
* This can be a problem if we want to print several columns of values, since they can vary in length.
* We will now look at how they can be formatted.
* We will use the **.format()** method to format values.
* The general sytax of the format method is:<br>

"{f1}{f2}{f3}".format(variable1, variable2, variable3), where
* {f1} is the formatting string for variable1
* {f2} is the formatting string for variable2
* {f3} is the formatting string for variable3
* The formatting string is in the format: :[flags][width][.precision]type, where <br>


Flag | Explanation
----------- | -------------------
< | The field will be left-aligned within the available space. This is usually the default for strings
> | The field will be right-aligned within the available space. This is the default for numbers.
0 | If the width field is preceded by a zero ('0') character, zero-padding for numeric types will be enabled.
, | Signals the use of a comma for a thousands separator
= | Forces the padding to be placed after the sign (if any) but before the digits
^ | Forces the field to be centered within the available space
space | Indicates that negative number should be prefixed with a minus sign and a space should be used for a positive number

* width - the total width of the field
* precision - the number of decimal digits. If this is used with a string, then it indicates that the string should be truncated to this number of positions

Type | Explanation
----------|--------------------
d | Decimal representation of signed integers
i | Decimal representation of signed integers
e | Floating point exponential format (lowercase).
E | Floating point exponential format (uppercase).
f | Floating point decimal format.
F |  Floating point decimal format.
g | Same as "e" if exponent is greater than -4 or less than precision, "f" otherwise.
G |  Same as "E" if exponent is greater than -4 or less than precision, "F" otherwise.
s |  String (converts any python object using str()).

In [14]:
from math import pi

# String Examples
print("{}".format("hello")) #|h|e|l|l|o|
print("{:8s}".format("hello")) #h|e|l|l|o| | | |
print("{:>8s}".format("hello")) #| | | |h|e|l|l|o|
print("{:^8s}".format("hello")) #| |h|e|l|l|o| | |
print("{:^4.3s}".format("hello")) #| |h|e| |

# Integer Examples
print("{:4d}".format(42))   #| | |4|2|
print("{:04d}".format(42)) #|0|0|4|2|

# Other Examples
one = "one"
two = "two"
three = "three"

print("{}{}{}".format("one","two","three"))
print("one = {:s}, two = {:s}, three = {:s}".format(one, two, three))
print("one = {:8s}, two = {:8s}, three = {:8s}".format(one, two, three))
print("one = {:.2s}, two = {:.2s}, three = {:.2s}".format(one, two, three))

print("x{:4.1f}x".format(pi))
print('{:,.2f}'.format(123456789.017))


Hi = 8.5
Lo = -4.4
print("{}{}".format(Hi, Lo))
print("Todays high is {}, the low is {}.".format(Hi, Lo))
print("Todays high is {:8.2f}, the low is {:8.2f}.".format(Hi, Lo))
print("{:^8s}{:^8s}".format("Hi","Low"))
print("{:8.1f}{:8.1f}".format(91.5, 66.4))
#print("{:<8.1f}{:<8.1f}".format(91.5, 66.4))
print("{:8.1f}{:8.1f}".format(101.5, 55.4))
print("{:8.1f}{:8.1f}".format(8.5, -4.4))




hello
hello   
   hello
 hello  
hel 
  42
0042
onetwothree
one = one, two = two, three = three
one = one     , two = two     , three = three   
one = on, two = tw, three = th
x 3.1x
123,456,789.02
8.5-4.4
Todays high is 8.5, the low is -4.4.
Todays high is     8.50, the low is    -4.40.
   Hi     Low   
    91.5    66.4
   101.5    55.4
     8.5    -4.4


# Further Reading from the Think Python Book
* [Assignment statements](https://greenteapress.com/thinkpython2/html/thinkpython2003.html#sec16)
* [Variable names](https://greenteapress.com/thinkpython2/html/thinkpython2003.html#sec17)
* [Expressions and statements](https://greenteapress.com/thinkpython2/html/thinkpython2003.html#sec18)
* [Script mode](https://greenteapress.com/thinkpython2/html/thinkpython2003.html#sec19)
* [Order of operations String operations Comments](https://greenteapress.com/thinkpython2/html/thinkpython2003.html#sec20)
* [String operations](https://greenteapress.com/thinkpython2/html/thinkpython2003.html#sec21)
* [Comments](https://greenteapress.com/thinkpython2/html/thinkpython2003.html#sec22)
* [Debugging](https://greenteapress.com/thinkpython2/html/thinkpython2003.html#sec23)
* [Glossary](https://greenteapress.com/thinkpython2/html/thinkpython2003.html#sec24)
* [Exercises](https://greenteapress.com/thinkpython2/html/thinkpython2003.html#sec25)

# Ex1: Last Digit
* Prompt for a number and print its last digit.
* Be sure to use to .format method to create your output.
* Use only math for your solution; do no use any string functions.
   
Enter a number: 23 <br>
Last Digit: 3 <br><br>


Enter a number: 5 <br>
Last Digit: 5 <br><br>


Enter a number: 10 <br>
Last Digit: 0 <br><br>

# Ex2: Two Digits
* Enter a two digit number and print out its digits separately.
*  Be sure to use to .format method to create your output.
* Use only math for your solution; do no use any string functions.

Enter a Number: 17 <br>
Ones Digit: 7 <br>
Tens Digit: 1 <br><br>


Enter a Number: 40 <br>
Ones Digit: 0 <br>
Tens Digit: 4 <br><br>


Enter a Number: 33 <br>
Ones Digit: 3 <br>
Tens Digit: 3 <br>

# Ex3: Swap Digits
* Enter a two digit number and swap their digits and create a new two digit number.
* Be sure to use the .format method to create your output.
* Use only math for your solution; do no use any string functions.

Enter a Number: 17 <br>
Swapped Number: 71 <br><br>


Enter a Number: 33 <br>
Swapped Number: 33 <br><br>


Enter a Number: 29 <br>
Swapped Number: 92

# Ex4: Last Two Digits
* Enter a number greater than zero and print the value its last two digits.
* Be sure to use to .format method to create your output. You will have to use leading zeros.
* Use only math for your solution; do no use any string functions.

Enter a Number: 178  <br>
Last Two Digits: 78  <br><br>

Enter a Number: 4 <br>
Last Two Digits: 04 <br><br>


Enter a Number: 100 <br>
Last Two Digits: 00

# Ex5: Tens Digit
* Enter a number and print its tens digit.
* Be sure to use the .format method to create your output
* Use only math for your solution; do no use any string functions.

Enter a Number: 72 <br>
Tens Digit: 7 <br><br>


Enter a Number: 1234 <br>
Tens Digit: 3 <br><br>


Enter a Number: 6 <br>
Tens Digit: 0

# Ex6: Sum of Digits
* Enter a three digit number and print the sum of the digits.
* Be sure to use the .format method to create your output.
* Use only math for your solution; do no use any string functions.

  
Enter a 3 Digit Number: 179 <br>
Sum of Digits: 17 <br><br>

Enter a 3 Digit Number: 829 <br>
Sum of Digits: 19 <br><br>

Enter a 3 Digit Number: 204 <br>
Sum of Digits: 6

# Ex7: Reverse Three Digits
* Enter a three digit number and reverse the digits.
* Be sure to use to .format method to create your output.
* Use only math for your solution; do no use any string functions.

Enter a 3 Digit Number: 291 <br>
Reverse of Digits: 192 <br><br>


Enter a 3 Digit Number: 321 <br>
Reverse of Digits: 123 <br><br>


Enter a 3 Digit Number: 444 <br>
Reverse of Digits: 444

# Ex8: Merge Two Numbers
* Prompt for two two-digit numbers.
* Merge them into single number as follows:
    * Tens digit from first number
    * Tens digit from second number
    * Units digit from first number
    * Units digit form second number
* Use only math for your solution; do no use any string functions.

Enter First 2 Digit Number: 12 <br>
Enter Second 2 Digit Number: 34 <br>
Merged Number: 1324 <br><br>


Enter First 2 Digit Number: 33 <br>
Enter Second 2 Digit Number: 55 <br>
Merged Number: 3535 <br><br>

Enter First 2 Digit Number: 29 <br>
Enter Second 2 Digit Number: 11 <br>
Merged Number: 2191

#Ex9: Fractional Part
* Enter a positive real number and print its fractional part.
* Use only math for your solution; do no use any string functions.
* **Hint**: Round the result to 10 digits using the round function.
  

Enter a Number: 17.9 <br>
Fractional Part: 0.9 <br><br>


Enter a Number: 10.34 <br>
Fractional Part: 0.34 <br><br>


Enter a Number: 0.001 <br>
Fractional Part: 0.001 <br>

# Ex10: First Digit After Decimal
* Prompt for a positive real number
* Print the first digit to the right of the decimal point.
* Use only math for your solution; do no use any string functions.

Enter Number: 1.79 <br>
Tenths Value: 7 <br><br>


Enter Number: 1.34 <br>
Tenths Value: 3 <br><br>

Enter Number: 0.001 <br>
Tenths Value: 0